swaps/swaps_service.go

```diff
@@ -125,254 +183,1057 @@ func (svc *swapsService) EnableAutoSwaps(ctx context.Context, lnClient lnclient.
+	fees := pairInfo.Fees
+	serviceFeePercentage := boltz.Percentage(fees.Percentage)
+
+	serviceFee := boltz.CalculatePercentage(serviceFeePercentage, amount)
+	networkFee := fees.MinerFees.Lockup + fees.MinerFees.Claim
+
+	logger.Logger.WithFields(logrus.Fields{
+		"serviceFee": serviceFee,
+		"networkFee": networkFee,
+	}).Info("Calculated fees for swap out")
+
+	albyFee := &boltz.ExtraFees{
+		Percentage: AlbySwapServiceFee,
+		Id:         "albyServiceFee",
+	}
+
+	dbSwap := db.Swap{
+		Type:               constants.SWAP_TYPE_OUT,
+		State:              constants.SWAP_STATE_PENDING,
+		SendAmount:         amount,
+		DestinationAddress: destination,
+		PaymentHash:        paymentHash,
+		Preimage:           hex.EncodeToString(preimage),
+		AutoSwap:           autoSwap,
+	}
+
+	var ourKeys *btcec.PrivateKey
+	var swap *boltz.CreateReverseSwapResponse
+
+	err = svc.db.Transaction(func(tx *gorm.DB) error {
+		err := tx.Save(&dbSwap).Error
+		if err != nil {
+			return err
+		}
+
+		defer func() {
+			if err != nil {
+				svc.markSwapState(&dbSwap, constants.SWAP_STATE_FAILED)
+			}
+		}()
+
+		ourKeys, err = svc.keys.GetSwapKey(dbSwap.ID)
+		if err != nil {
+			return fmt.Errorf("error generating swap child private key: %w", err)
+		}
+
+		swap, err = svc.boltzApi.CreateReverseSwap(boltz.CreateReverseSwapRequest{
+			From:           boltz.CurrencyBtc,
+			To:             boltz.CurrencyBtc,
+			ClaimPublicKey: ourKeys.PubKey().SerializeCompressed(),
+			PreimageHash:   preimageHash[:],
+			InvoiceAmount:  amount,
+			Description:    "Lightning to on-chain swap",
+			PairHash:       pairInfo.Hash,
+			ReferralId:     "alby",
+			ExtraFees:      albyFee,
+		})
+
+		if err != nil {
+			return fmt.Errorf("could not create swap: %s", err)
+		}
+
+		swapTreeJson, err := json.Marshal(swap.SwapTree)
+		if err != nil {
+			return err
+		}
+
+		err = tx.Model(&dbSwap).Updates(&db.Swap{
+			SwapId:             swap.Id,
+			Invoice:            swap.Invoice,
+			LockupAddress:      swap.LockupAddress,
+			TimeoutBlockHeight: swap.TimeoutBlockHeight,
+			BoltzPubkey:        hex.EncodeToString(swap.RefundPublicKey),
+			SwapTree:           datatypes.JSON(swapTreeJson),
+		}).Error
+		if err != nil {
+			return err
+		}
+
+		// commit transaction
+		return nil
+	})
+
+	if err != nil {
+		logger.Logger.WithError(err).WithFields(logrus.Fields{
+			"paymentHash": paymentHash,
+		}).Error("Failed to save swap")
+		return nil, err
+	}
+
+	logger.Logger.WithField("swapId", swap.Id).Info("Swap created")
+
+	boltzWs := svc.boltzApi.NewWebsocket()
+	if err := boltzWs.Connect(); err != nil {
+		svc.markSwapState(&dbSwap, constants.SWAP_STATE_FAILED)
+		return nil, fmt.Errorf("could not connect to Boltz websocket: %w", err)
+	}
+
+	if err := boltzWs.Subscribe([]string{swap.Id}); err != nil {
+		_ = boltzWs.Close()
+		svc.markSwapState(&dbSwap, constants.SWAP_STATE_FAILED)
+		return nil, err
+	}
+
+	go svc.startSwapOutListener(&dbSwap, boltzWs)
```

Comment:
I'm don't think we handle the websocket well in both these listener functions (swap out and in).

Shouldn't we handle reconnections? is there a recommended way to do this?

I see there is a `Connected()` method and a `Reconnect()` method available.

Reply:
Done, we keep trying till it connects before starting this go routine